elfo-core/src/message/repr.rs

```diff
@@ -0,0 +1,340 @@
+        ) -> Result<(), decode::Error> {
+            let data = decode::from_slice(buffer)?;
+            ptr::write(
+                out_ptr.cast::<MessageRepr<M>>().as_ptr(),
+                MessageRepr::new(data),
+            );
+            Ok(())
+        }
+
+        pub(super) unsafe fn write_msgpack<M: Message>(
+            ptr: NonNull<MessageRepr>,
+            out: &mut Vec<u8>,
+            limit: usize,
+        ) -> Result<(), encode::Error> {
+            let data = &ptr.cast::<MessageRepr<M>>().as_ref().data;
+            let mut out = LimitedWrite(out, limit);
+            encode::write_named(&mut out, data)
+        }
+    });
+}
+
+// === VTable registration & lookup ===
+
+/// A list of all registered message vtables via the `linkme` crate.
+/// Used only for collecting, all lookups are done via a hashmap.
+// Reexported in `elfo::_priv`.
+#[doc(hidden)]
+#[linkme::distributed_slice]
+pub static MESSAGE_VTABLES_LIST: [&'static MessageVTable] = [..];
+
+static MESSAGE_VTABLES_MAP: Lazy<FxHashMap<(&'static str, &'static str), &'static MessageVTable>> =
+    Lazy::new(|| {
+        MESSAGE_VTABLES_LIST
+            .iter()
+            .map(|vtable| ((vtable.protocol, vtable.name), *vtable))
+            .collect()
+    });
+
+impl MessageVTable {
+    /// Finds a vtable by protocol and name.
+    /// Used for deserialization of `AnyMessage` and in networking.
+    pub(crate) fn lookup(protocol: &str, name: &str) -> Option<&'static Self> {
+        // Extend lifetimes to static in order to get `(&'static str, &'static str)`.
+        // SAFETY: this pair doesn't overlive the function.
+        let (protocol, name) = unsafe {
+            (
+                std::mem::transmute::<&str, &'static str>(protocol),
+                std::mem::transmute::<&str, &'static str>(name),
+            )
+        };
+
+        MESSAGE_VTABLES_MAP.get(&(protocol, name)).copied()
+    }
+}
+
+/// Checks that all registered message have different protocol and name.
+/// Returns a list of duplicates if it's violated.
+pub(crate) fn check_uniqueness() -> Result<(), Vec<(String, String)>> {
+    if MESSAGE_VTABLES_MAP.len() == MESSAGE_VTABLES_LIST.len() {
+        return Ok(());
+    }
+
+    Err(MESSAGE_VTABLES_LIST
+        .iter()
+        .filter(|vtable| {
+            let stored = MessageVTable::lookup(vtable.protocol, vtable.name).unwrap();
+            MessageTypeId::new(stored) != MessageTypeId::new(vtable)
+        })
+        .map(|vtable| (vtable.protocol.to_string(), vtable.name.to_string()))
+        .collect::<FxHashSet<_>>()
+        .into_iter()
+        .collect::<Vec<_>>())
+}
+
+// === LimitedWrite ===
+
+cfg_network!({
+    use std::io;
+
+    struct LimitedWrite<W>(W, usize);
+
+    impl<W: io::Write> io::Write for LimitedWrite<W> {
+        #[inline]
+        fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+            if buf.len() > self.1 {
+                self.1 = 0;
+                return Ok(0);
```

Comment:
Is there a test for over-the-limit `.write_msgpack()`? I think it’s worthwhile to check that rmp-serde handles `Ok(0)` correctly.

Reply:
I don't remember any test for it, only manual once, good point

Reply:
It's already tested in elfo-network/codec. Anyway, more tests (under miri) are added here